components/script/dom/html/htmldialogelement.rs

```diff
@@ -142,6 +145,112 @@ impl HTMLDialogElement {
+
+        // Step 5. Remove subject's open attribute.
+        subject.remove_attribute(&ns!(), &local_name!("open"), can_gc);
+
+        // TODO: Step 6. If is modal of subject is true, then request an element to be removed from the top layer given subject.
+
+        // TODO: Step 7. Let wasModal be the value of subject's is modal flag.
+
+        // Step 8. Set is modal of subject to false.
+        self.is_modal.set(false);
+
+        // Step 9. If result is not null, then set subject's returnValue attribute to result.
+        if let Some(new_value) = result {
+            *self.return_value.borrow_mut() = new_value;
+        }
+
+        // TODO: Step 10. Set subject's request close return value to null.
+
+        // TODO: Step 11. Set subject's request close source element to null.
+
+        // TODO: Step 12. If subject's previously focused element is not null, then:
+        // TODO: Step 12.1. Let element be subject's previously focused element.
+        // TODO: Step 12.2. Set subject's previously focused element to null.
+        // TODO: Step 12.3. If subject's node document's focused area of the document's DOM anchor is a shadow-including inclusive descendant of subject, or wasModal is true, then run the focusing steps for element; the viewport should not be scrolled by doing this step.
+
+        // Step 13. Queue an element task on the user interaction task source given the subject element to fire an event named close at subject.
+        let target = self.upcast::<EventTarget>();
+        self.owner_global()
+            .task_manager()
+            .user_interaction_task_source()
+            .queue_simple_event(target, atom!("close"));
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#queue-a-dialog-toggle-event-task>
+    pub fn queue_dialog_toggle_event_task(&self, old_state: DOMString, new_state: DOMString, source: Option<DomRoot<Element>>) {
+        // TODO: Step 1. If element's dialog toggle task tracker is not null, then:
+        // TODO: Step 1.1. Set oldState to element's dialog toggle task tracker's old state.
+        // TODO: Step 1.2. Remove element's dialog toggle task tracker's task from its task queue.
+        // TODO: Step 1.3. Set element's dialog toggle task tracker to null.
+        // Step 2. Queue an element task given the DOM manipulation task source and element to run the following steps:
+        let this = Trusted::new(self);
+        let old_state = old_state.to_string();
+        let new_state = new_state.to_string();
+
+        let trusted_source = source.as_ref().map(|el| {
+            Trusted::new(el.upcast::<EventTarget>())
+        });
+
+        self.owner_global()
+            .task_manager()
+            .dom_manipulation_task_source()
+            .queue(task!(fire_toggle_event: move || {
+                let this = this.root();
+
+                let source = trusted_source.as_ref().map(|s| {
```

Comment:
This all feels very weird but was what I could come up with when fighting the compiler...

Reply:
he trait bound `Root<Dom<Element>>: DomObject` is not satisfied [E0277]
unsatisfied trait bound
Help: the trait `script_bindings::reflector::DomObject` is not implemented for `script_bindings::root::Root<script_bindings::root::Dom<dom::element::Element>>`

When I try to do it without the up and down casting I get this?